Failure to join tables with encrypted data - different number of partitions

When performing join on encrypted data (Opaque) with another dataframe - there is a rdd.zip exception when the two dataframes are not with the same number of partitions. 

Caused by: java.lang.IllegalArgumentException: Can't zip RDDs with unequal numbers of partitions: List(3, 1)
        at org.apache.spark.rdd.ZippedPartitionsBaseRDD.getPartitions(ZippedPartitionsRDD.scala:57)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:248)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:246)
        at scala.Option.getOrElse(Option.scala:121)
        at org.apache.spark.rdd.RDD.partitions(RDD.scala:246)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:1930)
        at org.apache.spark.rdd.RDD.count(RDD.scala:1134)
        at edu.berkeley.cs.rise.opaque.execution.ObliviousUnionExec$$anonfun$executeBlocked$3.apply$mcJ$sp(operators.scala:346)
        at edu.berkeley.cs.rise.opaque.execution.ObliviousUnionExec$$anonfun$executeBlocked$3.apply(operators.scala:346)
        at edu.berkeley.cs.rise.opaque.execution.ObliviousUnionExec$$anonfun$executeBlocked$3.apply(operators.scala:346)
        at edu.berkeley.cs.rise.opaque.Utils$.time(Utils.scala:93)
        at edu.berkeley.cs.rise.opaque.execution.ObliviousUnionExec.executeBlocked(operators.scala:346)
        at edu.berkeley.cs.rise.opaque.execution.EncryptedSortExec.executeBlocked(EncryptedSortExec.scala:33)
        at edu.berkeley.cs.rise.opaque.execution.EncryptedSortMergeJoinExec.executeBlocked(operators.scala:296)
        at edu.berkeley.cs.rise.opaque.execution.ObliviousProjectExec.executeBlocked(operators.scala:207)
        at edu.berkeley.cs.rise.opaque.execution.OpaqueOperatorExec$class.executeCollect(operators.scala:147)
        at edu.berkeley.cs.rise.opaque.execution.ObliviousProjectExec.executeCollect(operators.scala:200)
        at org.apache.spark.sql.Dataset$$anonfun$org$apache$spark$sql$Dataset$$execute$1$1.apply(Dataset.scala:2193)
        at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:57)
        at org.apache.spark.sql.Dataset.withNewExecutionId(Dataset.scala:2546)
        at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$execute$1(Dataset.scala:2192)
        at org.apache.spark.sql.Dataset$$anonfun$org$apache$spark$sql$Dataset$$collect$1.apply(Dataset.scala:2197)
        at org.apache.spark.sql.Dataset$$anonfun$org$apache$spark$sql$Dataset$$collect$1.apply(Dataset.scala:2197)
        at org.apache.spark.sql.Dataset.withCallback(Dataset.scala:2559)
        at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collect(Dataset.scala:2197)
        at org.apache.spark.sql.Dataset.collect(Dataset.scala:2173)
      

Also, Opaque rdd partitions count always returns 0.

Regular join (on non-encrypted data) works well on the same data set.

To reproduce perform a simple join on two dataframes with different partitions size.
for example, using the person dataset provided in previous join issue:
- Use person.json or any input json file with multiple lines
- Create a new json file with a subset of the above (few lines)
- place the two files in the same folder
- Using spark shell, load the files into df using *
- load the subset into another df
- perform join on the two dataframes using the PersonID col - the results is the subset (with each dataframe having different # of partitions)
- perform the same join on encrypted data - rdd.zip error

--- a/src/main/scala/edu/berkeley/cs/rise/opaque/execution/operators.scala
+++ b/src/main/scala/edu/berkeley/cs/rise/opaque/execution/operators.scala
@@ -327,8 +327,8 @@ case class ObliviousUnionExec(
     left.output
 
   override def executeBlocked() = {
-    val leftRDD = left.asInstanceOf[OpaqueOperatorExec].executeBlocked()
-    val rightRDD = right.asInstanceOf[OpaqueOperatorExec].executeBlocked()
+    var leftRDD = left.asInstanceOf[OpaqueOperatorExec].executeBlocked()
+    var rightRDD = right.asInstanceOf[OpaqueOperatorExec].executeBlocked()
     Utils.ensureCached(leftRDD)
     time("Force left child of ObliviousUnionExec") { leftRDD.count }
     Utils.ensureCached(rightRDD)
@@ -336,13 +336,18 @@ case class ObliviousUnionExec(
 
     // RA.initRA(leftRDD)
 
-    val unioned = leftRDD.zipPartitions(rightRDD) { (leftBlockIter, rightBlockIter) =>
-      (leftBlockIter.toSeq ++ rightBlockIter.toSeq) match {
-        case Seq(leftBlock, rightBlock) =>
-          Iterator(Utils.concatEncryptedBlocks(Seq(leftBlock, rightBlock)))
-        case Seq(block) => Iterator(block)
-        case Seq() => Iterator.empty
+    val num_left_partitions = leftRDD.getNumPartitions
+    val num_right_partitions = rightRDD.getNumPartitions
+    if (num_left_partitions != num_right_partitions) {
+      if (num_left_partitions > num_right_partitions) {
+        leftRDD = leftRDD.coalesce(num_right_partitions)
+      } else {
+        rightRDD = rightRDD.coalesce(num_left_partitions)
       }
+    }
+    val unioned = leftRDD.zipPartitions(rightRDD) {
+      (leftBlockIter, rightBlockIter) =>
+        Iterator(Utils.concatEncryptedBlocks(leftBlockIter.toSeq ++ rightBlockIter.toSeq))
     }
     Utils.ensureCached(unioned)
     time("ObliviousUnionExec") { unioned.count }

--- a/src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala
+++ b/src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala
@@ -334,4 +334,19 @@ class OpaqueMultiplePartitionSuite extends OpaqueOperatorTests {
     .getOrCreate()
 
   override def numPartitions = 3
+
+  import testImplicits._
+  def makePartitionedDF[A <: Product : scala.reflect.ClassTag : scala.reflect.runtime.universe.TypeTag](
+    data: Seq[A], securityLevel: SecurityLevel, numPartitions: Int, columnNames: String*): DataFrame =
+    securityLevel.applyTo(
+      spark.createDataFrame(
+        spark.sparkContext.makeRDD(data, numPartitions))
+        .toDF(columnNames: _*))
+  testAgainstSpark("join with different numbers of partitions (#34)") { securityLevel =>
+    val p_data = for (i <- 1 to 16) yield (i.toString, i * 10)
+    val f_data = for (i <- 1 to 256 - 16) yield ((i % 16).toString, (i * 10).toString, i.toFloat)
+    val p = makeDF(p_data, securityLevel, "pk", "x")
+    val f = makePartitionedDF(f_data, securityLevel, numPartitions + 1, "fk", "x", "y")
+    p.join(f, $"pk" === $"fk").collect.toSet
+  }
 }

